server/datastore/mysql/apple_mdm.go

```diff
@@ -1950,6 +1950,79 @@ func bulkDeleteHostDiskEncryptionKeysDB(ctx context.Context, tx sqlx.ExtContext,
 	return err
 }
 
+func (ds *Datastore) MDMAppleGetEULAMetadata(ctx context.Context) (*fleet.MDMAppleEULA, error) {
+	// Currently, there can only be one EULA in the database, so we pick
+	// the first entry we find.
+	stmt := "SELECT name, created_at, token FROM eulas LIMIT 1"
+	var eula fleet.MDMAppleEULA
+	if err := sqlx.GetContext(ctx, ds.reader, &eula, stmt); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ctxerr.Wrap(ctx, notFound("MDMAppleEULA"))
+		}
+		return nil, ctxerr.Wrap(ctx, err, "get EULA metadata")
+	}
+	return &eula, nil
+}
+
+func (ds *Datastore) MDMAppleGetEULABytes(ctx context.Context, token string) (*fleet.MDMAppleEULA, error) {
+	stmt := "SELECT name, bytes FROM eulas WHERE token = ?"
+	var eula fleet.MDMAppleEULA
+	if err := sqlx.GetContext(ctx, ds.reader, &eula, stmt, token); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ctxerr.Wrap(ctx, notFound("MDMAppleEULA"))
+		}
+		return nil, ctxerr.Wrap(ctx, err, "get EULA bytes")
+	}
+	return &eula, nil
+}
+
+func (ds *Datastore) MDMAppleInsertEULA(ctx context.Context, eula *fleet.MDMAppleEULA) error {
+	// check if we already have an EULA in the database. This is a bit odd,
+	// at the moment we only allow one global EULA, but I haven't find a
+	// good way to enforce this check at the database level.
+	_, err := ds.MDMAppleGetEULAMetadata(ctx)
+
+	// if we didn't get any errors, the EULA exists, return an alreadyExists error
+	if err == nil {
+		return alreadyExists("MDMAppleEULA", "")
+	}
+
+	// if we got any error that's not a "not found", return that error
+	// before moving forward
+	if !fleet.IsNotFound(err) {
+		return ctxerr.Wrap(ctx, err, "checking if EULA already exists")
+	}
+
+	stmt := `
+          INSERT INTO eulas (name, bytes, token)
+	  VALUES (?, ?, ?)
+	`
+
+	_, err = ds.writer.ExecContext(ctx, stmt, eula.Name, eula.Bytes, eula.Token)
```

Comment:
Yeah this is a tricky approach, as it is racy and made worse by reading from the read replica to check if it exists and writing to the primary. I think we should leverage the DB constraints and use a unique constraint column (maybe just an ID for now, and we always pass eg. "1", but in the future as Sarah mentions it could probably be a team/no team id). This way, we can just insert directly and if we get a unique constraint error we know it already exists, without a race.

Reply:
thanks! that makes a lot of sense!